### DISCUSSION TITLE
Use ~= operator in TestStore

### DISCUSSION THREAD

**Discussion by andreweades** [pointfreeco/swift-composable-architecture#1846] [Ideas]
I use UUID and other types of ID a lot and I often wish that I could maintain that without having to have a special test version. I feel that I'm not really testing things as "real world" as I would like for the sake of State being Equatable.

How about a TestEquatable protocol that uses **~=** or can be switched between **==** and **~=** then it would be trivial to separate out. At the moment I have to override **==** to ignore ID which I'm still a little unsure about whether that will come back to bite me.

The new decency stuff makes this less of a problem for UUID but I still find that I'm creating structs of "equal" things but with unique IDs in my code.

**Comment by mbrandonw** [comment]
Hi @andreweades, I'm not sure I fully understand. Are you saying you want to be able to use `UUID`s in your models, but not have to deal controlling that dependency, and so essentially remove the `id` from consideration when checking for equality?

If so, I would highly recommend against that. It is completely possible for two distinct values to be "equal", save for the `id`, and if you allow mixing the two up in tests you can let some very subtle bugs through. In fact, in this week's episode we demonstrated that (at around [22:22](https://www.pointfree.co/episodes/ep220-modern-swiftui-dependencies-testing-part-2#t1342)). We accidentally substituted "duplicating" logic where we meant to have "editing" logic, and that led to a test failure in which only the IDs differed.

Now, having said that, if you truly, truly do want to go down this road (and again I highly recommend against it 🙂), then it seems you could create a kind of `@AlwaysEqual` property wrapper that simply returns `true`. It could even check the `_XCTIsTesting` from XCTestDynamicOverlay under the hood if you only want to do that in tests.

Or, maybe I've completely misunderstand your concerns. If so, could you share some code to make things more concrete?
